File: app/src/main/java/com/monk/service/StartService.java
package com.monk.service;

import android.content.Intent;
import android.os.Binder;
import android.os.IBinder;

import com.monk.activity.base.BaseService;
import com.monk.commonutils.LogUtil;

import java.lang.ref.WeakReference;

/**
 * @author monk
 * @date 2018-12-21
 */
public class StartService extends BaseService {

    private DemoBinder demoBinder;


    @Override
    public IBinder onBind(Intent intent) {
        LogUtil.i(simpleName, "onBind() -- " + intent);
        demoBinder = new DemoBinder(mService);
        return demoBinder;
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        LogUtil.i(tag,simpleName+"：onStartCommand(Intent)");
        return START_STICKY;
    }

    public static class DemoBinder extends Binder implements IBindInterface {
        private final WeakReference<BaseService> weakReference;

        public DemoBinder(BaseService mService) {
            weakReference = new WeakReference<>(mService);
        }

        @Override
        public void execute() {

            LogUtil.i("DemoBinder","执行了 binder 里面方法 execute()");
        }

        /**
         * 获取 Binder 依赖的 Service，方便调用者使用 Service 里的方法
         *
         * @return
         */
        public BaseService getService() {
            if (weakReference.get() != null) {
                return weakReference.get();
            }
            return null;
        }
    }

    public interface IBindInterface {
        void execute();
    }
}

File: module_fragment/src/main/java/com/monk/modulefragment/service/StartService.kt
package com.monk.modulefragment.service

import android.content.Intent
import android.os.Binder
import android.os.IBinder
import com.monk.activity.base.BaseService
import com.monk.commonutils.LogUtil
import java.lang.ref.WeakReference

/**
 * @author monk
 * @date 2018-12-21
 */
class StartService : BaseService<StartService?>() {
    private var demoBinder: DemoBinder? = null
    override fun onBind(intent: Intent): IBinder? {
        LogUtil.i(simpleName, "onBind() -- $intent")
        demoBinder = DemoBinder(mService)
        return demoBinder
    }

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        LogUtil.i(tag, "$simpleName：onStartCommand(Intent)")
        return START_STICKY
    }

    class DemoBinder(mService: BaseService<*>?) : Binder(), IBindInterface {
        private val weakReference: WeakReference<BaseService<*>?> = WeakReference(mService)
        override fun execute() {
            LogUtil.i("DemoBinder", "执行了 binder 里面方法 execute()")
        }

        /**
         * 获取 Binder 依赖的 Service，方便调用者使用 Service 里的方法
         *
         * @return
         */
        val service: BaseService<*>?
            get() = if (weakReference.get() != null) {
                weakReference.get()
            } else null

    }

    interface IBindInterface {
        fun execute()
    }
}